### PR TITLE
refactor: move GetTransaction to node/transaction.cpp

### PR DESCRIPTION
**Pull request by theStack**
~This PR is based on #22383, which should be reviewed first~ (merged by now).

In [yesterday's PR review club session to PR 22383](https://bitcoincore.reviews/22383), the idea of moving the function `GetTransaction(...)` from src/validation.cpp to src/node/transaction.cpp came up. With this, the circular dependency "index/txindex -> validation -> index/txindex" is removed (see change in `lint-circular-dependencies.sh`). Thanks to jnewbery for suggesting and to sipa for providing historical background.

Relevant IRC log:
```
17:52 <jnewbery> Was anyone surprised that GetTransaction() is in validation.cpp? It seems to me that node/transaction.cpp would be a more appropriate place for it.
17:53 <raj_> jnewbery, +1
17:53 <stickies-v> agreed!
17:54 <glozow> jnewbery ya
17:54 <jnewbery> seems weird that validation would call into txindex. I wonder if we remove this function, then validation would no longer need to #include txindex
17:54 <sipa> GetTransaction predates node/transaction.cpp, and even the generic index framework itself :)
17:55 <sipa> (before 0.8, validation itself used the txindex)
17:55 <jnewbery> (and GetTransaction() seems like a natural sibling to BroadcastTransaction(), which is already in node/transaction.cpp)
17:55 <jnewbery> sipa: right, this is not meant as a criticism of course. Just wondering if we can organize things a bit more rationally now that we have better separation between things.
17:55 <sipa> jnewbery: sure, just providing background
17:56 <sipa> seems very reasonable to move it elsewhere now
```

The commit should be trivial to review with `--color-moved`.

### CODE REVIEW

File: src/node/transaction.h [hunk 1]
```diff
@@ -6,10 +6,13 @@
 #define BITCOIN_NODE_TRANSACTION_H
 
 #include <attributes.h>
+#include <consensus/params.h>
```
Comment (jnewbery) [inline src/node/transaction.h:9]: You don't need to include this. You can forward declare Consensus::Params:

```c++
namespace Consensus {
struct Params;
}
```

**Review by jnewbery — COMMENTED** [review]
Concept ACK. I have a branch that does exactly the same thing :)

File: src/node/transaction.h (hunk 1 above)
Comment (theStack) [reply src/node/transaction.h:9]: Ah, thanks! I was naively only `grep`ing for `Consensus::Params;` to look how other headers would forward-declare this (concluding that they don't and including the header is the preferred way), which is obviously not how it works :grimacing:

**Comment by theStack** [comment]
> Concept ACK. I have a branch that does exactly the same thing :)

My bad, should have communicated before to avoid double work. Feel free to open yours and I'll happily review :)

**Comment by jnewbery** [comment]
> My bad, should have communicated before to avoid double work. Feel free to open yours and I'll happily review :)

Not a problem at all! It's good to see that we separately arrived at the same answer :)

**Comment by LarryRuane** [comment]
concept ACK

File: test/lint/lint-circular-dependencies.sh [hunk 2]
```diff
@@ -10,7 +10,6 @@ export LC_ALL=C
 
 EXPECTED_CIRCULAR_DEPENDENCIES=(
     "chainparamsbase -> util/system -> chainparamsbase"
-    "index/txindex -> validation -> index/txindex"
```
Comment (promag) [inline test/lint/lint-circular-dependencies.sh:13]: 🎉

**Review by promag — COMMENTED** [review]
Code review ACK abc57e1f0882a1a2bb20474648419979af6e383d. With `git show --color-moved=dimmed_zebra` verified the move-only, `#includes` changes, added forward declarations and drop of circular dependency 🎉

**Comment by promag** [comment]
Happy to re-ACK with https://github.com/bitcoin/bitcoin/pull/22383#discussion_r674700841

**Comment by theStack** [comment]
Added a commit with jnewbery's [comment](https://github.com/bitcoin/bitcoin/pull/22383#discussion_r674700841)/[documentation](https://github.com/bitcoin/bitcoin/pull/22383#discussion_r674695881) follow-ups from #22383, [as suggested by MarcoFalke](https://github.com/bitcoin/bitcoin/pull/22383#issuecomment-885130926). Was not sure if the RPC doc change is also appropriate to put in this PR (since it was only concerned with the `GetTransaction` function in the beginning), OTOH it is called by the `getrawtransaction` RPC, so it should be okay.

**Comment by jnewbery** [comment]
Code review ACK f685a13bef0418663015ea6d8f448f075510c0ec

Thanks for doing this!

**Comment by rajarshimaitra** [comment]
tACK https://github.com/bitcoin/bitcoin/pull/22528/commits/f685a13bef0418663015ea6d8f448f075510c0ec

Had put it in my potential todos. @theStack  at the work amazingly fast.. 🔥

File: src/node/transaction.cpp [hunk 3]
```diff
@@ -104,3 +107,38 @@ TransactionError BroadcastTransaction(NodeContext& node, const CTransactionRef t
 
     return TransactionError::OK;
 }
+
+CTransactionRef GetTransaction(const CBlockIndex* const block_index, const CTxMemPool* const mempool, const uint256& hash, const Consensus::Params& consensusParams, uint256& hashBlock)
+{
+    LOCK(cs_main);
```
Comment (rajarshimaitra) [inline src/node/transaction.cpp:113]: Maybe it's also a good opportunity to address https://github.com/bitcoin/bitcoin/pull/22383#discussion_r675069128 here?

Comment (mjdietzx) [reply src/node/transaction.cpp:113]: +1

**Comment by mjdietzx** [comment]
crACK f685a13bef0418663015ea6d8f448f075510c0ec

File: src/node/transaction.cpp (hunk 3 above)
Comment (jnewbery) [reply src/node/transaction.cpp:113]: I'd say no. Better to keep this as move only and leave lock cleanups for a future PR. That makes this easier to review (and avoids invalidating the three ACKs on the current branch).

**Comment by practicalswift** [comment]
Concept ACK

Very nice to see the list of circular dependencies shrink :)

**Comment by LarryRuane** [comment]
Code review, test ACK f685a13bef0418663015ea6d8f448f075510c0ec

File: src/node/transaction.cpp (hunk 3 above)
Comment (theStack) [reply src/node/transaction.cpp:113]: The PR was planned to be one of the move-only refactoring kinds that should be trivial to review (strictly speaking, with the follow-up commit changing the RPC help text it's not a pure refactor anymore, but that I'd still consider acceptable). Changing locks is significantly harder to review, so I tend to agree with jnewbery here and leave it as it is by now.

**Comment by theStack** [comment]
Rebased on master (now that #22383 is merged, there is no need to include its commit anymore).

**Comment by MarcoFalke** [comment]
Please don't force push, as this requires re-review. This is a GitHub bug that doesn't affect us.

**Comment by theStack** [comment]
> Please don't force push, as this requires re-review. This is a GitHub bug that doesn't affect us.

Ah, sorry wasn't aware that displaying the first commit is happening due to a GitHub bug (if I understood you correctly). Force-pushed back to how it was before.

**Review by promag — COMMENTED** [review]
ACK f685a13bef0418663015ea6d8f448f075510c0ec. Only changes since last review are new comment about reorg and bip30 and improved help text.

**Pull request merged**